// @azure/cosmos/dist-esm/client/UserDefinedFunction/UserDefinedFunction.js
import { __awaiter } from "tslib";
import { createUserDefinedFunctionUri, getIdFromLink, getPathFromLink, isResourceValid, ResourceType } from "../../common";
import { UserDefinedFunctionResponse } from "./UserDefinedFunctionResponse";
/**
 * Used to read, replace, or delete a specified User Definied Function by id.
 *
 * @see {@link UserDefinedFunction} to create, upsert, query, read all User Defined Functions.
 */
export class UserDefinedFunction {
    /**
     * @hidden
     * @param container The parent {@link Container}.
     * @param id The id of the given {@link UserDefinedFunction}.
     */
    constructor(container, id, clientContext) {
        this.container = container;
        this.id = id;
        this.clientContext = clientContext;
    }
    /**
     * Returns a reference URL to the resource. Used for linking in Permissions.
     */
    get url() {
        return createUserDefinedFunctionUri(this.container.database.id, this.container.id, this.id);
    }
    /**
     * Read the {@link UserDefinedFunctionDefinition} for the given {@link UserDefinedFunction}.
     * @param options
     */
    read(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const path = getPathFromLink(this.url);
            const id = getIdFromLink(this.url);
            const response = yield this.clientContext.read({
                path,
                resourceType: ResourceType.udf,
                resourceId: id,
                options
            });
            return new UserDefinedFunctionResponse(response.result, response.headers, response.code, this);
        });
    }
    /**
     * Replace the given {@link UserDefinedFunction} with the specified {@link UserDefinedFunctionDefinition}.
     * @param body The specified {@link UserDefinedFunctionDefinition}.
     * @param options
     */
    replace(body, options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (body.body) {
                body.body = body.body.toString();
            }
            const err = {};
            if (!isResourceValid(body, err)) {
                throw err;
            }
            const path = getPathFromLink(this.url);
            const id = getIdFromLink(this.url);
            const response = yield this.clientContext.replace({
                body,
                path,
                resourceType: ResourceType.udf,
                resourceId: id,
                options
            });
            return new UserDefinedFunctionResponse(response.result, response.headers, response.code, this);
        });
    }
    /**
     * Delete the given {@link UserDefined}.
     * @param options
     */
    delete(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const path = getPathFromLink(this.url);
            const id = getIdFromLink(this.url);
            const response = yield this.clientContext.delete({
                path,
                resourceType: ResourceType.udf,
                resourceId: id,
                options
            });
            return new UserDefinedFunctionResponse(response.result, response.headers, response.code, this);
        });
    }
}
